src/java/org/apache/cassandra/index/sai/memory/MemtableIndex.java

```diff
@@ -61,32 +93,157 @@ public boolean isEmpty()
         writeCount.increment();
-        estimatedMemoryUsed.add(ram);
-        return ram;
     }
 
     public RangeIterator search(Expression expression, AbstractBounds<PartitionPosition> keyRange)
     {
-        return index.search(expression, keyRange);
+        RangeConcatIterator.Builder builder = RangeConcatIterator.builder();
+
+        for (MemoryIndex index : rangeIndexes)
+        {
+            builder.add(index.search(expression, keyRange));
+        }
+        return builder.build();
     }
 
-    public Iterator<Pair<ByteComparable, PrimaryKeys>> iterator()
+    /**
+     * NOTE: returned data may contain partition key not within the provided min and max which are only used to find
+     * corresponding subranges. We don't do filtering here to avoid unnecessary token comparison. In case of JBOD,
+     * min/max should align exactly at token boundaries. In case of tiered-storage, keys within min/max may not
+     * belong to the given sstable.
+     *
+     * @param min minimum partition key used to find min subrange
+     * @param max maximum partition key used to find max subrange
+     *
+     * @return iterator of indexed term to primary keys mapping in sorted by indexed term and primary key.
+     */
+    public Iterator<Pair<ByteComparable, Iterator<PrimaryKey>>> iterator(DecoratedKey min, DecoratedKey max)
     {
-        return index.iterator();
+        int minSubrange = min == null ? 0 : boundaries.getShardForKey(min);
+        int maxSubrange = max == null ? rangeIndexes.length - 1 : boundaries.getShardForKey(max);
+
+        List<Iterator<Pair<ByteComparable, PrimaryKeys>>> rangeLists = new ArrayList<>(maxSubrange - minSubrange + 1);
+        for (int i = minSubrange; i <= maxSubrange; i++)
+            rangeLists.add(rangeIndexes[i].iterator());
+
+        return MergeIterator.get(rangeLists, (o1, o2) -> ByteComparable.compare(o1.left, o2.left, ByteComparable.Version.OSS41),
+                                 new PrimaryKeysMergeReducer(rangeIndexes.length, clusteringComparator));
+    }
+
+    private static class PrimaryKeysMergeReducer extends MergeIterator.Reducer<Pair<ByteComparable, PrimaryKeys>, Pair<ByteComparable, Iterator<PrimaryKey>>>
+    {
+        private final Pair<ByteComparable, PrimaryKeys>[] toMerge;
+        private final int size;
+        private final Comparator<PrimaryKey> comparator;
+
+        @SuppressWarnings("unchecked")
+        PrimaryKeysMergeReducer(int size, ClusteringComparator clusteringComparator)
+        {
+            this.toMerge = new Pair[size];
+            this.size = size;
+            this.comparator = (o1, o2) -> {
+                if (!o1.partitionKey().equals(o2.partitionKey()))
+                    return o1.partitionKey().compareTo(o2.partitionKey());
+                return clusteringComparator.compare(o1.clustering(), o2.clustering());
+            };
+        }
+
+        @Override
+        public void reduce(int idx, Pair<ByteComparable, PrimaryKeys> current)
```

Comment:
I know it is not related to this ticket, but it is not very clear to me what the `idx` really means here.
The comment on the top-level abstract class doesn't help either. 
Maybe because you already know how that works, you could improve the docs on the abstract class so they are more clear? 

```java
      /**
         * combine this object with the previous ones.
         * intermediate state is up to your implementation.
         */
        public abstract void reduce(int idx, In current);
```

I mean this comment doesn't mention `idx` at all. What range ofr values is expected here? What does it really mean? 
I can see it was added by Sylvain Lebresne in `Storage engine refactor, a.k.a CASSANDRA-8099`, and I guess they forgot to update the comment. Because earlier, without the idx I can understand what it was supposed to do.

Reply:
I have added some comments on our implementation that hopefully make this a bit clearer. The `idx` directly relates to one of the `sources` that are passed into a `Merger`. None of this code is readable but does make sense when you debug it. As above I would prefer not to update these comments but to leave that for the CEP.